### PR TITLE
Dawid szpejna/cpp support

### PR DESCRIPTION
**Pull request by DawidSzpejna**
## Description

- Change syntax of specified attributes and qualifiers. 
- Add required functions, variables and macros. 
- Add language linkage.
- Change `atexit` implementation and whole destructors handling  

## Motivation and Context

Adapt libphoenix to build libstdc++.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [ ] Tested by hand on: (list targets here).
- [x] Covered by tests provided by GCC: 
  - `ia32-generic-qemu`
  - `armv7a9-zynq7000`

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [ ] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: stdlib/strtod.c
Comment (nalajcie) [inline stdlib/strtod.c:153]: shame it's almost ^C^V of the `strtod` but I don't have a clear solution on how to unify the implementations.

Was this implementation tested?

File: include/math.h
Comment (nalajcie) [inline include/math.h:36]: nitpicking: external `( )` are not needed for function aliases, please remove them

File: string/strerror.c
Comment (nalajcie) [inline string/strerror.c:73]: use `/*`

File: string/strerror.c
Comment (nalajcie) [inline string/strerror.c:75]: we could probably align `return`'s to make them more readable.

To silent the `clang-format` You would need to surround the code block with `/* clang-format off */` and `/* clang-format on */`

**Review by nalajcie — CHANGES_REQUESTED** [review]
Thanks for the changes, I've added some remarks (mostly about coding style) to be fixed before the merge ;).

**Comment by nalajcie** [comment]
also
 - please rebase on top of current `master:HEAD` to resolve conflicts in `pthread.h`
 - please fix small style issues reported by `clang-format` above

File: include/stdlib.h
Comment (DawidSzpejna) [reply include/stdlib.h]: I found a way how to avoid using `__cxa_atexit` in libstdc++. It requires to build gcc with special flag.

File: include/limits.h
Comment (ziemleszcz) [inline include/limits.h:19]: Nothing is actually changed in this header.

File: posix/stubs.c
Comment (DawidSzpejna) [reply posix/stubs.c:60]: I am not sure how the `setlocale` should behave if you pass `""` as a argument because we don't support the full implementation of this function.

Comment (ziemleszcz) [reply posix/stubs.c:60]: Set default locale - POSIX.

File: stdlib/atexit.c
Comment (DawidSzpejna) [reply stdlib/atexit.c:24]: I was taking this https://gcc.gnu.org/onlinedocs/gccint/Initialization.html into consideration when i was naming this variable. On the second line, this kind of functions are called *destructors* or *termination routines*.

File: stdlib/atexit.c
Comment (ziemleszcz) [reply stdlib/atexit.c]: My bad, <sys/threads.h> is needed because of mutex.

File: stdlib/atexit.c
Comment (ziemleszcz) [inline stdlib/atexit.c:32]: You can move it inside atexit_common.

File: stdlib/atexit.c
Comment (ziemleszcz) [inline stdlib/atexit.c:81]: It doesn't look good: there is no need to check `i` inside loop and restart the loop on new node allocation. I suggest using this scheme:

for (i = 0; i < ATEXIT_MAX; i++) {
  if (have free slot)
    break;
}
if (i == ATEXIT_MAX) {
   alloc  new node
   if (alloc fails)
     return;
  update  head and set i=0
}
assign function to slot i

File: posix/stubs.c
Comment (ziemleszcz) [inline posix/stubs.c:31]: Please remove underscores.

File: posix/stubs.c
Comment (ziemleszcz) [inline posix/stubs.c:33]: I suugest using one buffer here and doing sprintf(buf, C, POSIX) on each setlocal(*, NULL) call. User program can damage content of this buffer.

File: posix/stubs.c
Comment (ziemleszcz) [inline posix/stubs.c:65]: You can handle this case together with POSIX,

File: include/stdlib.h
Comment (ziemleszcz) [reply include/stdlib.h:36]: Remainder: please move `ATEXIT_TYPE` and `CXA_ATEXIT_TYPE` to `atexit.c`

File: stdlib/div.c
Comment (ziemleszcz) [reply stdlib/div.c:32]: it's correct but I would prefer using native types.

File: posix/stubs.c
Comment (DawidSzpejna) [reply posix/stubs.c]: I already improved the implementation of `setlocale`.

File: stdlib/atexit.c
Comment (DawidSzpejna) [reply stdlib/atexit.c:122]: At the moment it is possible to use more than 32 destructors. I use a linked list to store additional functions.

File: include/stdlib.h
Comment (DawidSzpejna) [reply include/stdlib.h:36]: I deleted `ATEXIT_TYPE` and `CXA_ATEXIT_TYPE`. Instead, I used bool values to indicate what function is registered. `0` if the function doesn't need arguments, '1' otherwise.

**Comment by nalajcie** [comment]
@agkaminski please scroll through the changes to know what's needed for cpp support in standard library

File: .codespell_ignore
Comment (agkaminski) [inline .codespell_ignore:1]: I don't know if I like global ignore for this - it's somewhat likely typo to be made. Maybe use:
- `n` and `d` or
- `num` and `den` or
- `numerator` and `denominator`
instead?

File: stdlib/atexit.c
Comment (DawidSzpejna) [reply stdlib/atexit.c:67]: According to POSIX, `atexit` doesn't have defined arrnos and should return non-zero value upon unsuccessful completion. In this case I am not sure whether setting errno to `ENOMEN` is appropriate...

File: stdlib/atexit.c
Comment (DawidSzpejna) [reply stdlib/atexit.c:109]: But if i free the `node`, i will lose the information about the ''next'' node on which i should work. Should I change the position of `atexit_common.head` to hold information about current node?

File: .codespell_ignore
Comment (DawidSzpejna) [reply .codespell_ignore:1]: Maybe you're right. I was trying to be consistent with Linux man-page where parameters are called `numer` and `denom`. Nevertheless i will change it.

File: stdlib/atexit.c
Comment (agkaminski) [reply stdlib/atexit.c:109]: You're freeing node (`last = node` in line 104).

Comment (agkaminski) [reply stdlib/atexit.c:109]: Both `node` and `last` are local variables that are lost after returning from the function

Comment (DawidSzpejna) [reply stdlib/atexit.c:109]: Could you explain it more precisely? IMO this solution won't work correctly.

**Review by agkaminski — APPROVED** [review]
LGTM, let's wait for nalajcie's review

File: stdlib/atexit.c
Comment (ziemleszcz) [inline stdlib/atexit.c:53]: Head pointer race possibility - please acquire the mutex before getting the head.

File: stdlib/atexit.c
Comment (ziemleszcz) [inline stdlib/atexit.c:63]: Could you merge these two comments into one: allocate new node if there are no free slots left?

File: stdlib/atexit.c
Comment (ziemleszcz) [reply stdlib/atexit.c]: Looks fine for me but breaks Phoenix coding convention (local variables defined before the code). Just add `node` initialization after `mutexLock`.

**Review by ziemleszcz — APPROVED** [review]
LGTM

**Review by nalajcie — APPROVED** [review]
not a careful review but I need to approve because I requested changes before :).

I see my previous issues were resolved so I'm ok with current state of things.

**Review by agkaminski — APPROVED** [review]

**Pull request merged**
